test/prober/await.go

```diff
@@ -0,0 +1,155 @@
+	"knative.dev/eventing/test/lib/duck"
+	"knative.dev/eventing/test/lib/resources"
+	"sync"
+	"time"
+)
+
+func (p *prober) waitForKServiceReady(name, namespace string) error {
+	meta := resources.NewMetaResource(name, namespace, &servingType)
+	return duck.WaitForResourceReady(p.client.Dynamic, meta)
+}
+
+func (p *prober) waitForKServiceScale(name, namespace string, satisfyScale func(*int32) bool) error {
+	metricName := fmt.Sprintf("waitForKServiceScale/%s/%s", namespace, name)
+	_, span := trace.StartSpan(context.Background(), metricName)
+	defer span.End()
+
+	return wait.PollImmediate(duck.Interval, duck.Timeout, func() (bool, error) {
+		serving := p.client.Dynamic.Resource(servicesCR).Namespace(namespace)
+		unstruct, err := serving.Get(name, metav1.GetOptions{})
+		return p.isScaledTo(satisfyScale, unstruct, namespace, err)
+	})
+}
+
+func (p *prober) isScaledTo(satisfyScale func(*int32) bool, un *unstructured.Unstructured, namespace string, err error) (bool, error) {
+	if k8serrors.IsNotFound(err) {
+		// Return false as we are not done yet.
+		// We swallow the error to keep on polling.
+		// It should only happen if we wait for the auto-created resources, like default Broker.
+		return false, nil
+	} else if err != nil {
+		// Return error to stop the polling.
+		return false, err
+	}
+
+	content := un.UnstructuredContent()
+	maybeStatus, ok := content["status"]
+	if !ok {
+		return false, nil
+	}
+	status := maybeStatus.(map[string]interface{})
+	maybeTraffic, ok := status["traffic"]
+	if !ok {
+		return false, nil
+	}
+	traffic := maybeTraffic.([]interface{})
+	if len(traffic) > 1 {
+		return false, fmt.Errorf("traffic shouldn't be split to more then 1 revision: %v", traffic)
+	}
+	if len(traffic) == 0 {
+		// continue to wait
+		return false, nil
+	}
+	firstTraffic := traffic[0].(map[string]interface{})
+	revisionName := firstTraffic["revisionName"].(string)
+	deploymentName := fmt.Sprintf("%s-deployment", revisionName)
+
+	var dep *appsv1.Deployment
+	dep, err = p.client.Kube.Kube.AppsV1().Deployments(namespace).
+		Get(deploymentName, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		// Return false as we are not done yet.
+		return false, nil
+	} else if err != nil {
+		// Return error to stop the polling.
+		return false, err
+	}
+	return satisfyScale(dep.Spec.Replicas), nil
+}
+
+func (p *prober) waitForTriggerReady(name, namespace string) error {
+	meta := resources.NewMetaResource(name, namespace, lib.TriggerTypeMeta)
+	return duck.WaitForResourceReady(p.client.Dynamic, meta)
+}
+
+func (p *prober) waitForPodReady(name, namespace string) error {
+	podType := &metav1.TypeMeta{
+		Kind:       "Pod",
+		APIVersion: "v1",
+	}
+	meta := resources.NewMetaResource(name, namespace, podType)
+	return duck.WaitForResourceReady(p.client.Dynamic, meta)
+}
+
+type namedAwait struct {
+	name    string
+	routine awaitRoutine
+}
+
+type awaitRoutine func() error
+
+var waits []*namedAwait
+
+func awaitAll(log *zap.SugaredLogger) {
```

Comment:
FYI. There's an errGroup construct in Golang which is there for catching errors in routing running in the group: https://github.com/golang/sync/blob/master/errgroup/errgroup.go#L18

Reply:
Good idea. I don't know if I can use it here, as I do some additional logging there.

Reply:
Done with a22418c6e22709b7f05dbdc6d938b6330c438ff7